src/librustdoc/clean/types.rs

```diff
@@ -749,12 +756,16 @@ crate enum ItemKind {
     MacroItem(Macro),
     ProcMacroItem(ProcMacro),
     PrimitiveItem(PrimitiveType),
-    AssocConstItem(Type, Option<ConstantKind>),
-    /// An associated item in a trait or trait impl.
+    /// A required associated constant in a trait definition.
+    TyAssocConstItem(Type),
+    /// An associated associated constant in a trait impl or a provided one in a trait definition.
+    AssocConstItem(Type, ConstantKind),
+    /// A required associated type in a trait definition.
     ///
     /// The bounds may be non-empty if there is a `where` clause.
-    /// The `Option<Type>` is the default concrete type (e.g. `trait Trait { type Target = usize; }`)
-    AssocTypeItem(Box<Generics>, Vec<GenericBound>, Option<Type>),
+    TyAssocTypeItem(Box<Generics>, Vec<GenericBound>),
+    /// An associated type in a trait impl or a provided one in a trait impl.
+    AssocTypeItem(Typedef, Vec<GenericBound>),
```

Comment:
I really like this "specialization". It makes it much more clear what it's actually for.